pallets/xyk/src/lib.rs

```diff
@@ -11,92 +8,285 @@ mod mock;
+                asset_id.into(),
+                T::Lookup::unlookup(to),
+                amount,
+            )
+            .map_err(|_| Error::<T>::NotEnoughTokens)
+        }
+
+        fn mint_lp(lp: AssetId<T>, amount: Balance<T>, to: T::AccountId) {
+            pallet_assets::Pallet::<T>::mint(
+                RawOrigin::Signed(T::DEXAddr::get()).into(),
+                lp.into(),
+                T::Lookup::unlookup(to),
+                amount,
+            )
+            .unwrap()
+        }
+
+        fn burn_lp(lp: AssetId<T>, amount: Balance<T>, from: T::AccountId) {
+            pallet_assets::Pallet::<T>::burn(
+                RawOrigin::Signed(T::DEXAddr::get()).into(),
+                lp.into(),
+                T::Lookup::unlookup(from),
+                amount,
+            )
+            .unwrap()
+        }
+
+        pub fn lp_token(x: AssetId<T>, y: AssetId<T>) -> AssetId<T> {
+            let (x, y) = if x > y { (y, x) } else { (x, y) };
+            AssetId::<T>::decode(&mut T::AssetIdHash::hash_of(&(x, y)).encode().as_ref()).unwrap()
+        }
+
+        pub fn _swap(
+            origin: T::AccountId,
+            (from_id, from): (AssetId<T>, Balance<T>),
+            to_id: AssetId<T>,
+        ) -> Result<Balance<T>, Error<T>> {
+            let lp = Self::lp_token(from_id, to_id);
+
+            LPPairs::<T>::mutate(lp, |p| {
+                let p = p.as_mut().ok_or(Error::<T>::NoLiquidity)?;
+
+                Self::transfer(origin.clone(), T::DEXAddr::get(), from_id, from)?;
+
+                // x * y = k
+                // (x + dx) * (y - dy) = k
+                // y - dy = x * y / (x + dx)
+                // dy = y - x * y / (x + dx)
+                let new_x = p.x.1 + from;
+                let new_y = (p.x.1 * p.y.1) / new_x;
+
+                let dy = p.y.1 - new_y;
+
+                Self::transfer(T::DEXAddr::get(), origin.clone(), to_id, dy)
+                    .expect("Should always work by because of x*y=k architecture");
+
+                p.x.1 = new_x;
+                p.y.1 = new_y;
+
+                Ok(dy)
+            })
+        }
+
+        fn _create_lp(
+            origin: T::AccountId,
+            (x_id, x): (AssetId<T>, Balance<T>),
+            (y_id, y): (AssetId<T>, Balance<T>),
+        ) -> Result<(AssetId<T>, Balance<T>), Error<T>> {
+            let lp = Self::lp_token(x_id, y_id);
+
+            ensure!(!LPPairs::<T>::contains_key(lp), Error::LiquidityAlreadyExists);
+
+            Self::transfer(origin.clone(), T::DEXAddr::get(), x_id, x)?;
+            Self::transfer(origin.clone(), T::DEXAddr::get(), y_id, y)?;
+
+            pallet_assets::Pallet::<T>::create(
+                RawOrigin::Signed(T::DEXAddr::get()).into(),
+                lp.into(),
+                T::Lookup::unlookup(T::DEXAddr::get()),
+                T::LpMinBalance::get(),
+            )
+            .expect("Should always be random. Assume no collisions");
```

Comment:
`Must` is a more appropriate word here. `Should` indicates that it might not always be true, see [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).